SingleQuoteFixer - fix single quote char

Slamdunk opened the pull request:
- [x] Add configuration for fixing strings containing single-quotes, disabled by default
- [x] Create integration test with https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3155

keradus commented on src/Fixer/LanguageConstruct/ClassKeywordRemoveFixer.php:
```diff
@@ -219,7 +219,7 @@ private function replaceClassKeyword(Tokens $tokens, $classIndex)
         $tokens->insertAt($classBeginIndex, new Token(array(
             T_CONSTANT_ENCAPSED_STRING,
-            "'".$this->makeClassFQN($classImport, $classString)."'",
+            '\''.$this->makeClassFQN($classImport, $classString).'\'',
```
:-1: for this changes

Slamdunk replied:
You mean only `"'"` or the entire PR?

keradus replied:
we follow Sf rule here:
use single-quoted string except of cases when using it would require extra escaping

Slamdunk replied:
If so, I'm going to make this configurable and disabled by default.

keradus replied:
for which fixer?

Slamdunk replied:
`SingleQuoteFixer`: I'll rebase this PR on master instead of 2.2, and create a configuration whether to fix or not double-quoted strings that contain single-quotes; disabled by default for Symfony and BC compatibility

Slamdunk commented:
I've rebased to master since it's a new functionality, added the new configuration, made it backward compatible, created integration tests with #3155 

As often happens to me, I'm confident with the code, but the name of the new configuration it's a bit raw: `strings_containing_single_quote_chars`

SpacePossum commented on src/Fixer/StringNotation/SingleQuoteFixer.php:
```diff
@@ -21,16 +24,30 @@
      */
     public function getDefinition()
     {
+        $codeSamble = <<<'EOF'
```
little typo in the var name

Slamdunk replied:
Thanks

SpacePossum commented on src/Fixer/StringNotation/SingleQuoteFixer.php:
```diff
@@ -56,14 +73,27 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
             if (
                 '"' === $content[0] &&
-                false === strpos($content, "'") &&
+                (false === strpos($content, "'") || true === $this->configuration['strings_containing_single_quote_chars']) &&
```
please swap the tests, the config check is cheaper than the `strpos`

Slamdunk replied:
Done

SpacePossum commented on src/Fixer/StringNotation/SingleQuoteFixer.php:
```diff
@@ -56,14 +73,27 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
             ) {
                 $content = substr($content, 1, -1);
-                $content = str_replace(['\\"', '\\$'], ['"', '$'], $content);
+                $content = str_replace(['\\"', '\\$', '\''], ['"', '$', '\\\''], $content);
```
isn't this a BC break, as it always does this change no matter what has been configured?

Slamdunk replied:
No because the code enters here only if the configuration is enabled.

keradus commented:
Thank you @Slamdunk.
